hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/types/HazelcastObjectType.java

```diff
@@ -113,9 +114,9 @@ public void addField(Field field) {
         fields.add(field);
     }
 
-    public void finalizeFields() {
-        fields = List.of(fields.toArray(Field[]::new));
-        super.computeDigest();
+    public void finalizeFields(Map<String, HazelcastObjectType> typeMap) {
+        typeMap.values().forEach(type -> type.fields = ((NonTraversableList<Field>) type.fields).toReadonlyList());
+        typeMap.values().forEach(type -> type.computeDigest());
```

Comment:
is it computed in 2 passes due to (potentially circular) references between the types?

Reply:
`finalizeFields()` is also responsible for computing the digest since this computation should be done after the fields are finalized, and it is an implementation detail that the callers don't need to know. However, there is a problem here: if you do `typeMap.values().forEach(HazelcastObjectType::finalizeFields)` where `finalizeFields()` is the following, `computeDigest()` will fail since it tries to traverse the nested fields before they are finalized.
```java
public void finalizeFields() {
    fields = ((NonTraversableList<Field>) fields).toReadonlyList());
    computeDigest();
}
```
As a solution, the finalization is done in two passes. The reason of expecting `typeMap` is to speed up the process. Otherwise, `HazelcastObjectType` will need to collect all nested types in a similar collection. We can stop expecting this map, so that it looks more like an instance method, but I think, in this case, the performance gain justifies the violation of encapsulation and separation of concerns. WDYT? Of course, one may argue that `computeDigest()` can be exposed to the user by following a similar justification. However, I think "the collection of all referenced types" is a more general concept than our "digest" field, so exposing the former makes more sense to me.

Alternatively, we can compute the digest when it is used  (in `equals()`, `hashCode()` and `getFullTypeString()`), but such an implementation wouldn't be easily maintainable as it may be broken silently in a future release of Calcite.

Reply:
that was my understanding. IMO worth adding a comment in the code.
I am fine with `typeMap` - for current use case it is sufficient and has smaller runtime overhead than other options.

Reply:
c1fe5fbd

Reply:
✔️